source/dealii/dealii_hierarchy_helpers.cc

```diff
@@ -70,16 +74,165 @@ DealIIHierarchyHelpers<dim, VectorType>::build_restrictor(
+            {
+              delta_eig[k] =
+                  delta_eigenvector_matrix->el(row, dof_indices_map[k]) +
+                  _eigenvector_matrix->el(row, dof_indices_map[k]) /
+                      eigenvalues[row];
+            }
+          }
+          else
+          {
+            for (unsigned int k = 0; k < n_elem; ++k)
+            {
+              delta_eig[k] =
+                  delta_eigenvector_matrix->el(row, dof_indices_map[k]);
+            }
+          }
+
+          // Perform the matrix-vector multiplication
+          dealii::Vector<ScalarType> correction(n_elem);
+          agglomerate_system_matrix.vmult(correction, delta_eig);
+
+          // We would like to fill the delta correction matrix but we can't
+          // because we don't know the sparsity pattern. So we accumulate all
+          // the values and then fill the matrix using the set() function.
+          for (unsigned int k = 0; k < n_elem; ++k)
+            delta_correction_acc[std::make_pair(row, dof_indices_map[k])] +=
+                correction[k];
+        }
+      }
+
+      is_halo_agglomerate = true;
+    }
+
+    // Fill _delta_correction_matrix
+    for (auto const &entry : delta_correction_acc)
+      _delta_correction_matrix->set(entry.first.first, entry.first.second,
+                                    entry.second);
+    _delta_correction_matrix->compress(dealii::VectorOperation::insert);
+  }
+  else
+  {
+    _amge->setup_restrictor(agglomerate_params, n_eigenvectors, tolerance,
+                            *dealii_mesh_evaluator,
+                            locally_relevant_global_diag, *restrictor_matrix);
+    // If we don't use fast_ap, we don't need to keep _amge around. So release
+    // the memory.
+    _amge.reset();
+  }
 
   std::shared_ptr<Operator<VectorType>> op(
       new DealIITrilinosMatrixOperator<VectorType>(restrictor_matrix));
 
   return op;
 }
 
+template <int dim, typename VectorType>
+std::shared_ptr<Operator<VectorType>>
+DealIIHierarchyHelpers<dim, VectorType>::fast_multiply_transpose()
+{
+  Epetra_CrsMatrix *ap = nullptr;
+  // We want to use functions that have been deprecated in deal.II but they
+  // won't be removed in the foreseeable
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+  Epetra_Map range_map = _eigenvector_matrix->domain_partitioner();
+  Epetra_Map domain_map = _eigenvector_matrix->range_partitioner();
+#pragma GCC diagnostic pop
+
+  bool const transpose = true;
+  int error_code = EpetraExt::MatrixMatrix::Add(
+      _eigenvector_matrix->trilinos_matrix(), transpose, 1.,
+      _delta_correction_matrix->trilinos_matrix(), transpose, 1., ap);
+  ap->FillComplete(domain_map, range_map);
+
+  ASSERT(error_code == 0, "Problem when adding matrices");
+
+  // Copy the Epetra_CrsMatrix to a dealii::TrilinosWrappers::SparseMatrix
+  auto dealii_ap = std::make_shared<dealii::TrilinosWrappers::SparseMatrix>();
+  dealii_ap->reinit(*ap);
+  delete ap;
```

Comment:
I guess you can deduce it since the sparsity pattern is the union of the two others but neither graph is included in the other. I'll show you the math.